### PR TITLE
Support RegExp search and replace in file edit.

### PR DESCRIPTION
**Pull request by Jocs**
| Q                | A
| ---------------- | ---
| New feature?     | yes
| Fixed tickets    | #356 
| License          | MIT

### Description

- [x] Search string
- [x] Case sensitive
- [x] Whole word match
- [x] Use RegExp to search

屏幕快照 2019-10-03 上午12 33 36

---
This change is [Reviewable](https://reviewable.io/reviews/marktext/marktext/1422)

### CODE REVIEW

File: src/muya/lib/contentState/searchCtrl.js [hunk 1]
```diff
@@ -1,26 +1,51 @@
...
-  highlightIndex: -1
+import execall from 'execall'
+import { defaultSearchOption } from '../config'
+
+const matchString = (text, value, options) => {
+  const { isCaseSensitive, isWholeWord, isRegexp } = options
+  /* eslint-disable no-useless-escape */
+  const SPECIAL_CHAR_REG = /[\[\]\\^$.\|\?\*\+\(\)\/]{1}/g
+  /* eslint-enable no-useless-escape */
+  let SEARCH_REG = null
```
Comment (fxha) [inline src/muya/lib/contentState/searchCtrl.js]: Please use camel case instead upper case or what's the idea behind this?

File: src/muya/lib/contentState/searchCtrl.js [hunk 2]
```diff
@@ -1,26 +1,51 @@
...
+  if (isWholeWord) {
+    regStr = `\\b${regStr}\\b`
+  }
+
+  try {
+    // Add try catch expression because not all string can generate a valid RegExp. for example `\`.
+    SEARCH_REG = new RegExp(regStr, flag)
+    return execall(SEARCH_REG, text)
+  } catch (err) {
+    return []
```
Comment (fxha) [inline src/muya/lib/contentState/searchCtrl.js]: Should we notify the user about the issue?

Comment (Jocs) [reply src/muya/lib/contentState/searchCtrl.js]: I also think about this, I'll add the notification.

File: src/muya/lib/contentState/searchCtrl.js (hunk 1 above)
Comment (Jocs) [reply src/muya/lib/contentState/searchCtrl.js]: We also use this style to declare RegExp in other files, like `src/muya/lib/config.js`. Just for consistency

File: src/renderer/components/search/index.vue [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+        try {
+          const SEARCH_REG = new RegExp(searchValue)
+          if (searchValue && SEARCH_REG.test('')) {
+            throw new Error()
+          }
+          this.searchErrorMsg = ''
+        } catch (err) {
+          this.searchErrorMsg = `The regexp: /${searchValue}/ match too more.`
+          return
+        }
```
Comment (Jocs) [inline src/renderer/components/search/index.vue]: @fxha FYI, I add error notification here.

Comment (fxha) [reply src/renderer/components/search/index.vue]: I prefer `too much` instead `too more`.

Comment (Jocs) [reply src/renderer/components/search/index.vue]: Have changed to `RegExp: /${searchValue}/ match empty string.`.

**Comment by Jocs** [comment]
@fxha 
>I did some testing and found an issue when selecting a word, pressing Ctrl+F and then the searcher don't find any matches. If I write the selected word or paste it into the search bar all works fine.

fixed.

**Pull request merged**